### ISSUE TITLE
Incorrect number of nodes been provisioned with the batch fleet fix (#2130)

### ISSUE THREAD

**Issue by walkley** [aws/karpenter#2164]
## Version

**Karpenter**: image `public.ecr.aws/karpenter/controller:20220719` which included the commit for #2130.

**Kubernetes**: v1.21

## Expected Behavior

Karpenter should provision **500** nodes when I created deployment with 1500 pods.

## Actual Behavior

Karpenter provisioned **459** nodes in the first test, and **651** nodes in the second test.

## Steps to Reproduce the Problem

Follow the steps in [Getting Started with eksctl](https://karpenter.sh/v0.13.2/getting-started/getting-started-with-eksctl/) to create EKS cluster and deploy Karpenter, except some changes:
1. Change CPU limit, image tag and batchMaxDuration when run the helm install command:
```
  --set controller.resources.limits.cpu=2 \
  --set controller.image=public.ecr.aws/karpenter/controller:20220719 \
  --set controller.batchMaxDuration=20s
```
2. Set Provisioner:
```
apiVersion: karpenter.sh/v1alpha5
kind: Provisioner
metadata:
  name: default
spec:
  limits:
    resources:
      cpu: 2k
  provider:
    subnetSelector:
      karpenter.sh/discovery: ${CLUSTER_NAME}
    securityGroupSelector:
      karpenter.sh/discovery: ${CLUSTER_NAME}
  requirements:
    - key: node.kubernetes.io/instance-type
      operator: In
      values:
        - m5.large
        - m5d.large
    - key: kubernetes.io/arch
      operator: In
      values:
        - amd64
    - key: karpenter.sh/capacity-type
      operator: In
      values:
        - spot
        - on-demand
  ttlSecondsAfterEmpty: 120
```
3. Deploy 1500 pods:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: inflate
  namespace: default
spec:
  replicas: 1500
  selector:
    matchLabels:
      app: inflate
  template:
    metadata:
      labels:
        app: inflate
    spec:
      containers:
        - name: inflate
          image: public.ecr.aws/eks-distro/kubernetes/pause:3.2
          resources:
            requests:
              cpu: 512m
      nodeSelector:
        karpenter.sh/provisioner-name: default
```
4. Watch # of nodes been provisioned
```
kubectl get no -ojson | jq -r '.items | length'
```

## Resource Specs and Logs

1. Karpenter controller logs in first test which provisioned 459 nodes
[karpenter-6f7985bd74-nspj6.log](https://github.com/aws/karpenter/files/9160384/karpenter-6f7985bd74-nspj6.log)
    also, there was one CreateFleet call in CloudTrail:
    ```
    ...
    "eventTime": "2022-07-21T12:47:58Z",
    "eventSource": "ec2.amazonaws.com",
    "eventName": "CreateFleet",
    "awsRegion": "ap-northeast-1",
    "sourceIPAddress": "....",
    "userAgent": "aws-sdk-go/1.44.46 (go1.18.4; linux; amd64) karpenter.sh-7a6702d",
    "requestParameters": {
        "CreateFleetRequest": {
            "TargetCapacitySpecification": {
                "DefaultTargetCapacityType": "spot",
                "TotalTargetCapacity": 459
            },
    ...
    ```
2. Karpenter controller logs in second test which provisioned 651 node
[karpenter-6f7985bd74-xbkw5.log](https://github.com/aws/karpenter/files/9160395/karpenter-6f7985bd74-xbkw5.log)
    there was 4 CreateFleet calls in CloudTrail:
    ```
    ...
    "eventTime": "2022-07-21T13:17:10Z",
    "eventSource": "ec2.amazonaws.com",
    "eventName": "CreateFleet",
    "awsRegion": "ap-northeast-1",
    "sourceIPAddress": "...",
    "userAgent": "aws-sdk-go/1.44.46 (go1.18.4; linux; amd64) karpenter.sh-7a6702d",
    "requestParameters": {
        "CreateFleetRequest": {
            "TargetCapacitySpecification": {
                "DefaultTargetCapacityType": "spot",
                "TotalTargetCapacity": 500
            },
    ...
     "eventTime": "2022-07-21T13:17:56Z",
    "eventSource": "ec2.amazonaws.com",
    "eventName": "CreateFleet",
    "awsRegion": "ap-northeast-1",
    "sourceIPAddress": "...",
    "userAgent": "aws-sdk-go/1.44.46 (go1.18.4; linux; amd64) karpenter.sh-7a6702d",
    "requestParameters": {
        "CreateFleetRequest": {
            "TargetCapacitySpecification": {
                "DefaultTargetCapacityType": "spot",
                "TotalTargetCapacity": 203
            },
    ...
    "eventTime": "2022-07-21T13:17:59Z",
    "eventSource": "ec2.amazonaws.com",
    "eventName": "CreateFleet",
    "awsRegion": "ap-northeast-1",
    "sourceIPAddress": "...",
    "userAgent": "aws-sdk-go/1.44.46 (go1.18.4; linux; amd64) karpenter.sh-7a6702d",
    "requestParameters": {
        "CreateFleetRequest": {
            "TargetCapacitySpecification": {
                "DefaultTargetCapacityType": "spot",
                "TotalTargetCapacity": 2
            },
    ...
    "eventTime": "2022-07-21T13:18:08Z",
    "eventSource": "ec2.amazonaws.com",
    "eventName": "CreateFleet",
    "awsRegion": "ap-northeast-1",
    "sourceIPAddress": "...",
    "userAgent": "aws-sdk-go/1.44.46 (go1.18.4; linux; amd64) karpenter.sh-7a6702d",
    "requestParameters": {
        "CreateFleetRequest": {
            "TargetCapacitySpecification": {
                "DefaultTargetCapacityType": "spot",
                "TotalTargetCapacity": 72
            },
    ....
    ```

**Comment by walkley** [comment]
I can reproduce the second case(over-provisioning) with 30 pods(expect 10 nodes, but got 19), however, need minor code change as following. I will explain why introduced this change in the second section of this comment.
https://github.com/walkley/karpenter/commit/2b8caf21abd716f4b624caf36db3978736cf548a

### Steps to reproduce the issue with above debug build:
1. Build new container images with above changed code.
2. Deploy helm charts with new built images:
  ```
  helm upgrade --create-namespace --install karpenter karpenter/karpenter --namespace karpenter \
    --set serviceAccount.annotations.eks\\.amazonaws\\.com/role-arn=${KARPENTER_IAM_ROLE_ARN} \
    --set clusterName=${CLUSTER_NAME} \
    --set clusterEndpoint=${CLUSTER_ENDPOINT} \
    --set aws.defaultInstanceProfile=KarpenterNodeInstanceProfile-${CLUSTER_NAME} \
    --set controller.image=public.ecr.aws/whe/karpenter/controller@sha256:a524557fcbb9e5956720085175c14d08bbace6f3dcb333fd23dc0b92079eed28 \
    --set webhook.image=public.ecr.aws/whe/karpenter/webhook@sha256:513ac7c9fb957e3bc4a560496079dceab9bf7b03d60ea9ab07efa1eaabfae621
  ```
3. Provisioner:
  ```
  apiVersion: karpenter.sh/v1alpha5
  kind: Provisioner
  metadata:
    name: default
  spec:
    limits:
      resources:
        cpu: 2k
    provider:
      subnetSelector:
        karpenter.sh/discovery: ${CLUSTER_NAME}
      securityGroupSelector:
        karpenter.sh/discovery: ${CLUSTER_NAME}
    requirements:
      - key: node.kubernetes.io/instance-type
        operator: In
        values:
          - m5.large
          - m5d.large
      - key: kubernetes.io/arch
        operator: In
        values:
          - amd64
      - key: karpenter.sh/capacity-type
        operator: In
        values:
          - spot
          - on-demand
    ttlSecondsAfterEmpty: 120
  ```
4. Deploy 30 pods:
  ```
  apiVersion: apps/v1
  kind: Deployment
  metadata:
    name: inflate
    namespace: default
  spec:
    replicas: 30
    selector:
      matchLabels:
        app: inflate
    template:
      metadata:
        labels:
          app: inflate
      spec:
        containers:
          - name: inflate
            image: public.ecr.aws/eks-distro/kubernetes/pause:3.2
            resources:
              requests:
                cpu: 512m
        nodeSelector:
          karpenter.sh/provisioner-name: default
  ```
5. Karpenter provisioned **19**(10 + 9) new nodes:
  ```
  ...
  2022-07-24T04:54:54.387Z	INFO	controller.provisioning	Found 30 provisionable pod(s)	{"commit": "2b8caf2"}
  2022-07-24T04:54:54.387Z	INFO	controller.provisioning	Computed 10 new node(s) will fit 30 pod(s)	{"commit": "2b8caf2"}
  ...
  2022-07-24T04:54:57.895Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "2b8caf2"}
  ...
  2022-07-24T04:54:59.493Z	INFO	controller.provisioning	Found 30 provisionable pod(s)	{"commit": "2b8caf2"}
  2022-07-24T04:54:59.493Z	INFO	controller.provisioning	Computed 9 new node(s) will fit 27 pod(s)	{"commit": "2b8caf2"}
  2022-07-24T04:54:59.493Z	INFO	controller.provisioning	Computed 1 unready node(s) will fit 3 pod(s)	{"commit": "2b8caf2"}
  ```

### Root cause analysis
After reading the code and log I understood the mechanism of how Karpenter provision new nodes for pending pods, there are two related threads:
1. **Node reconciler:** maintain the list of nodes registered in cluster by watching node events from k8s API server and add/update/delete node in the cluster object.
2. **Provisioner:** provision new nodes for pending pods periodically, it will first check existing and in-flight nodes in above node list, and provision new nodes if necessary. The process of provisioning new nodes include launching new EC2 instances and registering these launched news nodes in K8s API server.

Node reconciler and Provisioner are work independently in two threads, they worked well in most cases:
```
Node reconciler  | watching node events.........................| received new nodes: 1,2,3...10 | watching again...
Provisioner:     | no inflight nodes, launch and register nodes | idle ...................................... | got 10 inflight nodes which can fit 30 pending pods | idle ...
```
However, if k8s API server is busy, or there are too many nodes in registering, node reconciler may take long time to receive all new nodes:
```
Node reconciler  | watching node events.........................| received new nodes: 1,....2,....3,....................10 | watching again...
Provisioner:     | no inflight nodes, launch and register nodes | idle ...................................... | got 4 inflight nodes which cannot fit 30 pending pods, launching new nodes... | idle ...
```

So it's clear why large number of nodes(500) and slow response time in node reconciler([simulate with 1s sleep in my debug build](https://github.com/walkley/karpenter/commit/2b8caf21abd716f4b624caf36db3978736cf548a)) will get over-provisioned nodes.

### Workaround
It worked if I set [batchMaxDuration](https://karpenter.sh/v0.13.2/tasks/configuration/#batchmaxduration) to 30s or larger number.

### Suggestion
I'm not sure it's feasible to add a way to synchronize the provisioner and node reconciler, to make sure all the nodes launched and registered by provisioner will be received by node reconciler in one provisioning loop.

**Comment by walkley** [comment]
I've tested the fix #2182 with several rounds of provisioning 500 nodes, it worked well!
Just noticed a few nodes get over-provisioned, the reason is that these nodes became unhealthy then back ready.

**Comment by walkley** [comment]
I am testing v0.14.0-rc.0, but still seeing a few over-provisioned nodes when running the case of 500 new nodes. To make this issue easier to reproduce, again I built [a debug version](https://github.com/walkley/karpenter/commit/93123dafa751f492d8353d0af149a51a3e23e2bb) with more logs added, and the key change of [sleeping 10s before creating inflight nodes](https://github.com/walkley/karpenter/commit/93123dafa751f492d8353d0af149a51a3e23e2bb#diff-1d10f18acbf8ccd4b28af2b0632dde4dc3d6ce363156984de4505447a9bb2ab9R80).

Here is the log when deploying 30 pods(same configuration as [previous comment](https://github.com/aws/karpenter/issues/2164#issuecomment-1193258514)) with this debug build, expected **10** nodes but actually got **20** nodes:
```
2022-07-31T09:13:34.069Z	DEBUG	controller.provisioning	Begin scheduling	{"commit": "c984cfa"}
2022-07-31T09:13:34.737Z	DEBUG	controller.provisioning	Discovered 434 EC2 instance types	{"commit": "c984cfa"}
2022-07-31T09:13:34.958Z	DEBUG	controller.provisioning	Discovered subnets: [subnet-08b4e545761b73088 (ap-northeast-1c) subnet-01904b5e14e8d313d (ap-northeast-1c) subnet-07c37106a7aa17619 (ap-northeast-1a) subnet-0066884add24ad0d2 (ap-northeast-1d) subnet-0333d26df904630f3 (ap-northeast-1a) subnet-00a83d5a154ac64ee (ap-northeast-1d)]	{"commit": "c984cfa"}
2022-07-31T09:13:35.047Z	DEBUG	controller.provisioning	Discovered EC2 instance types zonal offerings	{"commit": "c984cfa"}
2022-07-31T09:13:45.208Z	DEBUG	controller.provisioning	Creating in-flight nodes	{"commit": "c984cfa"}
2022-07-31T09:13:45.251Z	INFO	controller.provisioning	Found 30 provisionable pod(s)	{"commit": "c984cfa"}
2022-07-31T09:13:45.252Z	INFO	controller.provisioning	Computed 10 new node(s) will fit 30 pod(s)	{"commit": "c984cfa"}
...
2022-07-31T09:14:49.077Z	DEBUG	controller.provisioning	Begin scheduling	{"commit": "c984cfa"}
2022-07-31T09:14:50.840Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-183-93.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-rg5cs"}
2022-07-31T09:14:51.842Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-177-34.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-hl7qv"}
2022-07-31T09:14:52.842Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-166-71.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-gmtkh"}
2022-07-31T09:14:53.847Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-176-156.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-bnh2m"}
2022-07-31T09:14:53.847Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-166-122.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-ghx5r"}
2022-07-31T09:14:53.849Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-177-251.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-954t4"}
2022-07-31T09:14:53.849Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-170-221.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-wvqfx"}
2022-07-31T09:14:53.852Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-175-52.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-cxfvz"}
2022-07-31T09:14:53.853Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-177-34.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-md6cq"}
2022-07-31T09:14:53.853Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-174-110.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-vd5pj"}
2022-07-31T09:14:53.854Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-174-110.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-5v7cm"}
2022-07-31T09:14:53.854Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-166-122.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-v8lm7"}
2022-07-31T09:14:53.856Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-175-52.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-2rvlf"}
2022-07-31T09:14:53.857Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-170-221.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-426fd"}
2022-07-31T09:14:53.858Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-166-71.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-2jg5k"}
2022-07-31T09:14:53.858Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-183-93.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-dm6dv"}
2022-07-31T09:14:53.859Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-177-251.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-wdxhl"}
2022-07-31T09:14:53.859Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-176-156.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-76n9n"}
2022-07-31T09:14:53.859Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-175-52.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-nzq47"}
2022-07-31T09:14:53.861Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-177-251.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-jdr9c"}
2022-07-31T09:14:53.861Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-183-93.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-6bp2c"}
2022-07-31T09:14:53.861Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-170-221.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-2k9cq"}
2022-07-31T09:14:53.862Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-174-110.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-r9h5t"}
2022-07-31T09:14:53.862Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-166-122.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-jxwg7"}
2022-07-31T09:14:53.862Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-177-34.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-9rqhq"}
2022-07-31T09:14:53.863Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-176-156.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-qkkxn"}
2022-07-31T09:14:54.843Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-166-71.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-nm82d"}
2022-07-31T09:14:54.844Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-164-150.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-nv8q2"}
2022-07-31T09:14:54.845Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-164-150.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-h8nrq"}
2022-07-31T09:14:54.845Z	DEBUG	controller.pod-state	Bind to node: ip-192-168-164-150.ap-northeast-1.compute.internal	{"commit": "c984cfa", "pod": "default/inflate-6755dd8cc8-vnddr"}
2022-07-31T09:14:59.109Z	DEBUG	controller.provisioning	Creating in-flight nodes	{"commit": "c984cfa"}
2022-07-31T09:14:59.139Z	INFO	controller.provisioning	Found 30 provisionable pod(s)	{"commit": "c984cfa"}
2022-07-31T09:14:59.139Z	INFO	controller.provisioning	Computed 10 new node(s) will fit 30 pod(s)	{"commit": "c984cfa"}
...
```

I think the reason is again the consistency issue, this time it's between provisioner and pod-state controller:
```
pod-state.controller:  pod & node binding event | update node resource usage | another pod & node binding event | ...
Provisioner.Provision: Get pending pods | New Scheduler to create inflight nodes | Scheduler.Resolve to assign pending pods to inflight or new nodes
```
When creating inflight node, its available resource will be copied from node object in cluster. Since the time after getting pending pods and before creating inflight nodes is short, in most cases, the state of pending pods and available resource in node is consistent, so everything works well.

However, there still have pod binding event and node resource get updated between getting pending pods and creating inflight nodes in Provisioner, if provisioner running slower for example:
```
pod-state controller:  pod & node binding event | update node resource usage and other | another pod & node binding event | ...
Provisioner.Provision: Get pending pods ....................................................... | New Scheduler to create inflight nodes | ...
```

So I think it's necessary to make the updated node resource being synced after creating inflight nodes.

**Comment by walkley** [comment]
I've again tested the 500 nodes provisioning case based on the fix #2242 (controller image: `public.ecr.aws/karpenter/controller:90762392358efb6a01658de4aae0342c2030ac18`), in the 10 rounds of testings, it  always got exact 500 nodes been provisioned in less than 2 minutes. 

Nice job, thank you!
